users/vieting/experiments/switchboard/hybrid/feat/specaug_sorted.py

```diff
@@ -0,0 +1,173 @@
+    idxs = tf.expand_dims(tf.range(0, dim), 0)  # (1,dim)
+    pos_bc = tf.expand_dims(pos, 1)  # (batch,1)
+    pos2_bc = tf.expand_dims(pos2, 1)  # (batch,1)
+    cond = tf.logical_and(tf.greater_equal(idxs, pos_bc), tf.less(idxs, pos2_bc))  # (batch,dim)
+    if batch_axis > axis:
+        cond = tf.transpose(cond)  # (dim,batch)
+    cond = tf.reshape(cond, [tf.shape(x)[i] if i in (batch_axis, axis) else 1 for i in range(ndim)])
+    if sorted_indices is not None:
+        inverse_permutation = tf.argsort(sorted_indices)
+        cond = tf.gather(cond, inverse_permutation, axis=axis)
+    cond = tf.reshape(cond, [tf.shape(x)[i] if i in (batch_axis, axis) else 1 for i in range(ndim)])
+    from TFUtil import where_bc
+
+    x = where_bc(cond, 0.0, x)
+    return x
+
+
+def _random_mask(x, batch_axis, axis, min_num, max_num, max_dims, sorted_indices=None):
+    """
+    :param tf.Tensor x: (batch,time,feature)
+    :param int batch_axis:
+    :param int axis:
+    :param int|tf.Tensor min_num:
+    :param int|tf.Tensor max_num: inclusive
+    :param int|tf.Tensor max_dims: inclusive
+    :param tf.Tensor|None sorted_indices: sorted indices (optional)
+    """
+    from returnn.tf.compat import v1 as tf
+
+    n_batch = tf.shape(x)[batch_axis]
+    if isinstance(min_num, int) and isinstance(max_num, int) and min_num == max_num:
+        num = min_num
+    else:
+        num = tf.random_uniform(shape=(n_batch,), minval=min_num, maxval=max_num + 1, dtype=tf.int32)
+    # https://github.com/tensorflow/tensorflow/issues/9260
+    # https://timvieira.github.io/blog/post/2014/08/01/gumbel-max-trick-and-weighted-reservoir-sampling/
+    z = -tf.log(-tf.log(tf.random_uniform((n_batch, tf.shape(x)[axis]), 0, 1)))
+    _, indices = tf.nn.top_k(z, num if isinstance(num, int) else tf.reduce_max(num))
+    if isinstance(num, int):
+        for i in range(num):
+            x = _mask(
+                x,
+                batch_axis=batch_axis,
+                axis=axis,
+                pos=indices[:, i],
+                max_amount=max_dims,
+                sorted_indices=sorted_indices,
+            )
+    else:
+        _, x = tf.while_loop(
+            cond=lambda i, _: tf.less(i, tf.reduce_max(num)),
+            body=lambda i, x: (
+                i + 1,
+                tf.where(
+                    tf.less(i, num),
+                    _mask(
+                        x,
+                        batch_axis=batch_axis,
+                        axis=axis,
+                        pos=indices[:, i],
+                        max_amount=max_dims,
+                        sorted_indices=sorted_indices,
+                    ),
+                    x,
+                ),
+            ),
+            loop_vars=(0, x),
+        )
+    return x
+
+
+def specaugment_eval_func(data, network, time_factor=1):
+    x = data.placeholder
+    from returnn.tf.compat import v1 as tf
+
+    filter_layer = network.layers["features"].output.placeholder
```

Comment:
This is not the right layer, is it? It should be "conv_h_filter", I guess. I'm surprised that this even works. Did you test it?